save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/storage/AbstractMigrationStorage.kt

```diff
@@ -0,0 +1,100 @@
+package com.saveourtool.save.storage
+
+import com.saveourtool.save.utils.getLogger
+import com.saveourtool.save.utils.info
+import com.saveourtool.save.utils.warn
+import org.slf4j.Logger
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import java.nio.ByteBuffer
+import java.time.Instant
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.withLock
+
+/**
+ * Abstract storage which has an init method to migrate keys from old storage to new one
+ */
+abstract class AbstractMigrationStorage<O : Any, N : Any>(
+    private val oldStorage: Storage<O>,
+    private val newStorage: Storage<N>,
+) : Storage<O> {
+    private val log: Logger = getLogger(this.javaClass)
+    private val migrationLock = ReentrantReadWriteLock()
+    private var isMigrated = false
+
+    /**
+     * Init method which copies file from one storage to another
+     */
+    fun migrate() {
+        migrationLock.writeLock().withLock {
+            require(!isMigrated) {
+                "Migration cannot be called more than 1 time"
+            }
+            doMigrate()
+            isMigrated = true
+        }
+    }
+
+    private fun doMigrate() {
+        oldStorage.list()
+            .map { oldKey ->
+                oldKey to oldKey.toNewKey()
+            }
+            .filterWhen { (_, newKey) ->
+                newStorage.doesExist(newKey).map { !it }
```

Comment:
Any debug info here, maybe?

Reply:
added debug info when key already exists in new storage